distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso

```diff
@@ -559,10 +561,50 @@ type Table
     @group_by Widget_Helpers.make_column_name_vector_selector
     @order_by Widget_Helpers.make_order_by_selector
     add_row_number : Text -> Integer -> Integer -> Vector (Text | Integer | Column_Selector) | Text | Integer -> Vector (Text | Sort_Column) | Text -> Problem_Behavior -> Table
-    add_row_number self name="Row" from=1 step=1 group_by=[] order_by=[] on_problems=Problem_Behavior.Report_Warning =
-        _ = [name, from, step, group_by, order_by, on_problems]
-        msg = "`Table.add_row_number` is not yet implemented in the Database backend."
-        Error.throw (Unsupported_Database_Operation.Error msg)
+    add_row_number self (name:Text = "Row") (from:Integer = 1) (step:Integer = 1) group_by=[] order_by=[] on_problems=Problem_Behavior.Report_Warning =
+        problem_builder = Problem_Builder.new error_on_missing_columns=True
+        grouping_columns = self.columns_helper.select_columns_helper group_by True problem_builder
+        grouping_columns.each internal_column->
+            column = self.make_column internal_column
+            if column.value_type.is_floating_point then
+                problem_builder.report_other_warning (Floating_Point_Equality.Error column.name)
+        ordering = Table_Helpers.resolve_order_by self.columns order_by problem_builder
+        problem_builder.attach_problems_before on_problems <|
+            order_descriptors = case ordering.is_empty of
```

Comment:
Would it be possible to call `self.order_by` and then recursively call `add_row_number`?

Reply:
That is allowed yes.

Note that
```
t1 = t0.add_row_number order_by=["X"]
t2 = t0.order_by ["X"] . add_row_number
```
has slightly different semantics.

If `t0` is sorted according to some column `Y`, then in `t1` resulting rows will still be in the `Y` order, just the added column numbering will be according to `X`; while in `t2` the rows will be re-ordered to `X`.

Reply:
That is slightly database dependent - t1 can be reordered if no order by clause was specified.
Postgres will for example (as more efficient for it).
If multiple functions in SQL with order by clauses it tends to pick the first one I think.
Obv table order by dominates column.

Reply:
> That is slightly database dependent - t1 can be reordered if no order by clause was specified.

That's what I meant by:

> If `t0` is sorted according to some column `Y`

i.e. I assumed that `t0` has `ORDER BY Y`. And my point was that `add_row_number` itself will _not override_ this ordering, whereas an `order_by` followed by `add_row_number` will change it.